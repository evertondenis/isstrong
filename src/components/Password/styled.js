import styled from 'styled-components'

export const ContainerValidation = styled.div`
  display: block;
  width: 100%;
  font-size: 16px;
  color: #696d8c;

  .pwd,
  .confirm-pwd {
    margin: 10px 0;
    padding: 10px;
    width: 100%;
    background: #fff;
    border: 1px solid #b6b9d0;
    box-sizing: border-box;
    box-shadow: inset 0px 3px 3px rgba(0, 0, 0, 0.05);
    font-size: 1rem;
    color: #696d8c;
    outline: none;
    transition: all .2s linear;

    &:focus {
      outline: none;
    }
  }

  .isEmpty {
    border: 1px solid #b6b9d0;
  }

  .isWeak {
    border: 1px solid #f79682;
  }

  .isMedium {
    border: 1px solid #f7bc1c;
  }

  .isStrong {
    border: 1px solid #1fe6a8;
  }
`

export const StyledStages = styled.div`
  display: grid;
  grid-template-columns: repeat(3, 1fr);
  grid-gap: 5px;
  margin-bottom: 16px;

  .stage {
    border-radius: 5px;
    height: 8px;
    text-indent: -1000em;
    background-color: #eaeaf4;
    transition: all .2s linear;
  }

  .progress {
    width: 0;
    border-radius: 5px;
    height: 8px;
    background-color: #eaeaf4;
    opacity: 0;
    transition: all .2s linear;
  }

  &.isEmpty {
    border: 0;
  }

  &.isWeak {
    border: 0;
    .stage:first-child {
      .progress {
        width: 100%;
        opacity: 1;
        background-color: #f79682;
      }
    }
  }

  &.isMedium {
    border: 0;
    .stage:not(:last-child) {
      .progress {
        width: 100%;
        opacity: 1;
        background-color: #f7bc1c;
      }
    }
  }

  &.isStrong {
    border: 0;
    .stage {
      .progress {
        width: 100%;
        opacity: 1;
        background-color: #1fe6a8;
      }
    }
  }
`

export const StyledConfirm = styled.div`
  margin-top: 25px;

  .is-invalid {
    border-color: #f79682;
  }
`