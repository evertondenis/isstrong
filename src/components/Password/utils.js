export const passwordLength = (data, min = 6, max = 16) => {
  if (data.current && data.current.value) {
    const value = data.current.value
    const conditionLength = new RegExp(`^[\\w\\W]{${min},${max}}$`)
    const hasTotalCharacters = value.match(conditionLength)
    return hasTotalCharacters ? 'valid' : 'invalid'
  }
  return 'empty'
}

export const passwordNumber = (data) => {
  if (data.current && data.current.value) {
    const value = data.current.value
    const hasNumber = value.match(/\d/)
    return hasNumber ? 'valid' : 'invalid'
  }
  return 'empty'
}

export const passwordCapitalLetter = (data) => {
  if (data.current && data.current.value) {
    const value = data.current.value
    const hasCapitalLetter = value.match(/[A-Z]/)
    return hasCapitalLetter ? 'valid' : 'invalid'
  }
  return 'empty'
}

export const passwordSpecial = (data) => {
  if (data.current && data.current.value) {
    const value = data.current.value
    const hasSpecial = value.match(/\W/)
    return hasSpecial ? 'valid' : 'invalid'
  }
  return 'empty'
}


export const statusInput = (data) => {
  if (data) {
    const status = Object.keys(data)
    .map(step => data[step])
    .filter(item => item === 'valid' )

    return {
      1: 'isWeak',
      2: 'isMedium',
      3: 'isStrong'
    }[status.length] || 'isEmpty'
  }
  return 'isEmpty'
}