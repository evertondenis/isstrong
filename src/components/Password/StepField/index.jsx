import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import { Field } from './styled'

const Circle = ({ className, color }) => (
  <svg className={className} width="10" height="10" viewBox="0 0 10 10" fill="none" xmlns="http://www.w3.org/2000/svg">
    <circle cx="5" cy="5" r="5" fill={color}/>
  </svg>
)

class StepField extends PureComponent {
  render() {
    const { value, children } = this.props
    const status = {
      valid: '#1fe6a8',
      invalid : '#f79682'
    }[value] || '#eaeaf4'

    return (
      <Field>
        <Circle className="icon" color={status} />
        {children}
      </Field>
    )
  }
}

StepField.defaultProps = {
  children: ''
}

StepField.propTypes = {
  children: PropTypes.node,
  value: PropTypes.string.isRequired
}

export default StepField
