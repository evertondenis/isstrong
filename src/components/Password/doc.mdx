---
name: Password
menu: Components
---

import { Playground, PropsTable } from 'docz'
import Password from './'

# Password

<PropsTable of={Password} />

## Basic usage

<Playground>
  <Password confirmPwd />
</Playground>
