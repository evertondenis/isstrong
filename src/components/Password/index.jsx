import React, { Component, Fragment } from 'react'
import PropTypes from 'prop-types'
import { passwordLength, passwordCapitalLetter, passwordNumber, statusInput } from './utils'
import StepField from './StepField'
// import "font-awesome/css/font-awesome.css"
import { ContainerValidation, StyledStages, StyledConfirm } from './styled'

class Password extends Component {
  state = {
    field: '',
    status: {
      pwdLength: '',
      pwdCapital: '',
      pwdNumber: '',
      pwdEqual: ''
    }
  }

  inputRef = React.createRef()
  confirmRef = React.createRef()

  onChange = () => {
    const { min, max } = this.props
    this.setState({
      field: this.inputRef.current.value,
      status: {
        pwdLength: passwordLength(this.inputRef, min, max),
        pwdCapital: passwordCapitalLetter(this.inputRef),
        pwdNumber: passwordNumber(this.inputRef),
      }
    }, this.checkFields)
  }

  checkFields = () => {
    const { passwordValid, confirmPwd } = this.props
    const { field } = this.state

    if (confirmPwd) {
      const confirm = this.confirmRef.current.value
      const pwdEqual = confirm === field ? '' : 'is-invalid'
      this.setState({ pwdEqual })
      return confirm === field
        ? this.validPassword()
        : passwordValid(false, field)
    }
    return this.validPassword()
  }

  validPassword = () => {
    const { passwordValid } = this.props
    const { field } = this.state
    return statusInput(this.state.status) === 'isStrong'
      ? passwordValid(true, field)
      : passwordValid(false, field)
  }

  render() {
    const { min, confirmPwd, placeholder } = this.props
    const { status, pwdEqual } = this.state
    const { pwdLength, pwdCapital, pwdNumber } = status
    const styleStatus = statusInput(status)

    return (
      <Fragment>
        <ContainerValidation>
          <label>Senha</label>
          <input
            className={`pwd ${styleStatus}`}
            ref={this.inputRef}
            type="password"
            onChange={() => this.onChange()}
            placeholder={placeholder}
            required
          />
          <StyledStages className={`${styleStatus}`}>
            <div className="stage">
              <div className="progress">stage 1</div>
            </div>
            <div className="stage">
              <div className="progress">stage 2</div>
            </div>
            <div className="stage">
              <div className="progress">stage 3</div>
            </div>
          </StyledStages>
          <StepField value={pwdLength} > Pelo menos {min} caracteres</StepField>
          <StepField value={pwdCapital} > Pelo menos 1 letra maiúscula</StepField>
          <StepField value={pwdNumber} > Pelo menos 1 número</StepField>

          {confirmPwd && (
            <StyledConfirm>
              <label>Confirme sua senha</label>
              <input
                className={`confirm-pwd ${pwdEqual}`}
                ref={this.confirmRef}
                type="password"
                onChange={() => this.checkFields()}
                required
              />
            </StyledConfirm>
          )}
        </ContainerValidation>
      </Fragment>
    )
  }
}

Password.defaultProps = {
  confirmPwd: false,
  min: 6,
  max: 16,
  placeholder: ''
}

Password.propTypes = {
  confirmPwd: PropTypes.bool,
  min: PropTypes.number,
  max: PropTypes.number,
  placeholder: PropTypes.string,
  passwordValid: PropTypes.func
}

export default Password
