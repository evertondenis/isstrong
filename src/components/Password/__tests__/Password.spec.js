import React from 'react'
import { shallow, mount } from 'enzyme'
import 'jest-styled-components'
import Password from '../'
import 'setupTest'

describe('Password.', () => {
  it('Should be render <Password />.', () => {
    const min = 2
    const max = 6
    const wrapper = shallow(
      <Password min={min} max={max} />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('should pass a input value to the onChange handler', () => {
    const min = 2
    const max = 6
    const wrapper = mount(
      <Password min={min} max={max} passwordValid={(valid, pwd) => null} />
    )
    expect(wrapper.find('.pwd').length).toEqual(1)
    wrapper.find('.pwd').simulate('change')
    expect(wrapper.find('.pwd').length).toEqual(1)
  })

  it('should pass a input value to confirm handler', () => {
    const min = 2
    const max = 6
    const wrapper = mount(
      <Password min={min} max={max} confirmPwd passwordValid={(valid, pwd) => null} />
    )
    expect(wrapper.find('.pwd').length).toEqual(1)
    wrapper.find('.pwd').simulate('change')
    expect(wrapper.find('.pwd').length).toEqual(1)

    expect(wrapper.find('.confirm-pwd').length).toEqual(1)
    wrapper.find('.confirm-pwd').simulate('change')
    expect(wrapper.find('.confirm-pwd').length).toEqual(1)
  })
})
