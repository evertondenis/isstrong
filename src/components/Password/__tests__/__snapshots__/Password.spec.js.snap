// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Password. Should be render <Password />. 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Password
    confirmPwd={false}
    max={6}
    min={2}
    placeholder=""
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <styled.div>
        <label>
                Senha
        </label>
        <input
                className="pwd isEmpty"
                onChange={[Function]}
                placeholder=""
                required={true}
                type="password"
        />
        <styled.div
                className="isEmpty"
        >
                <div
                        className="stage"
                >
                        <div
                                className="progress"
                        >
                                stage 1
                        </div>
                </div>
                <div
                        className="stage"
                >
                        <div
                                className="progress"
                        >
                                stage 2
                        </div>
                </div>
                <div
                        className="stage"
                >
                        <div
                                className="progress"
                        >
                                stage 3
                        </div>
                </div>
        </styled.div>
        <StepField
                value=""
        >
                 Pelo menos 
                2
                 caracteres
        </StepField>
        <StepField
                value=""
        >
                 Pelo menos 1 letra maiúscula
        </StepField>
        <StepField
                value=""
        >
                 Pelo menos 1 número
        </StepField>
</styled.div>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <label>
            Senha
</label>,
          <input
            className="pwd isEmpty"
            onChange={[Function]}
            placeholder=""
            required={true}
            type="password"
/>,
          <styled.div
            className="isEmpty"
>
            <div
                        className="stage"
            >
                        <div
                                    className="progress"
                        >
                                    stage 1
                        </div>
            </div>
            <div
                        className="stage"
            >
                        <div
                                    className="progress"
                        >
                                    stage 2
                        </div>
            </div>
            <div
                        className="stage"
            >
                        <div
                                    className="progress"
                        >
                                    stage 3
                        </div>
            </div>
</styled.div>,
          <StepField
            value=""
>
             Pelo menos 
            2
             caracteres
</StepField>,
          <StepField
            value=""
>
             Pelo menos 1 letra maiúscula
</StepField>,
          <StepField
            value=""
>
             Pelo menos 1 número
</StepField>,
          false,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Senha",
          },
          "ref": null,
          "rendered": "Senha",
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "pwd isEmpty",
            "onChange": [Function],
            "placeholder": "",
            "required": true,
            "type": "password",
          },
          "ref": Object {
            "current": null,
          },
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <div
                className="stage"
>
                <div
                                className="progress"
                >
                                stage 1
                </div>
</div>,
              <div
                className="stage"
>
                <div
                                className="progress"
                >
                                stage 2
                </div>
</div>,
              <div
                className="stage"
>
                <div
                                className="progress"
                >
                                stage 3
                </div>
</div>,
            ],
            "className": "isEmpty",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <div
                  className="progress"
>
                  stage 1
</div>,
                "className": "stage",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "stage 1",
                  "className": "progress",
                },
                "ref": null,
                "rendered": "stage 1",
                "type": "div",
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <div
                  className="progress"
>
                  stage 2
</div>,
                "className": "stage",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "stage 2",
                  "className": "progress",
                },
                "ref": null,
                "rendered": "stage 2",
                "type": "div",
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <div
                  className="progress"
>
                  stage 3
</div>,
                "className": "stage",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "stage 3",
                  "className": "progress",
                },
                "ref": null,
                "rendered": "stage 3",
                "type": "div",
              },
              "type": "div",
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              " Pelo menos ",
              2,
              " caracteres",
            ],
            "value": "",
          },
          "ref": null,
          "rendered": Array [
            " Pelo menos ",
            2,
            " caracteres",
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": " Pelo menos 1 letra maiúscula",
            "value": "",
          },
          "ref": null,
          "rendered": " Pelo menos 1 letra maiúscula",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": " Pelo menos 1 número",
            "value": "",
          },
          "ref": null,
          "rendered": " Pelo menos 1 número",
          "type": [Function],
        },
        false,
      ],
      "type": [Function],
    },
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <styled.div>
          <label>
                    Senha
          </label>
          <input
                    className="pwd isEmpty"
                    onChange={[Function]}
                    placeholder=""
                    required={true}
                    type="password"
          />
          <styled.div
                    className="isEmpty"
          >
                    <div
                              className="stage"
                    >
                              <div
                                        className="progress"
                              >
                                        stage 1
                              </div>
                    </div>
                    <div
                              className="stage"
                    >
                              <div
                                        className="progress"
                              >
                                        stage 2
                              </div>
                    </div>
                    <div
                              className="stage"
                    >
                              <div
                                        className="progress"
                              >
                                        stage 3
                              </div>
                    </div>
          </styled.div>
          <StepField
                    value=""
          >
                     Pelo menos 
                    2
                     caracteres
          </StepField>
          <StepField
                    value=""
          >
                     Pelo menos 1 letra maiúscula
          </StepField>
          <StepField
                    value=""
          >
                     Pelo menos 1 número
          </StepField>
</styled.div>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <label>
              Senha
</label>,
            <input
              className="pwd isEmpty"
              onChange={[Function]}
              placeholder=""
              required={true}
              type="password"
/>,
            <styled.div
              className="isEmpty"
>
              <div
                            className="stage"
              >
                            <div
                                          className="progress"
                            >
                                          stage 1
                            </div>
              </div>
              <div
                            className="stage"
              >
                            <div
                                          className="progress"
                            >
                                          stage 2
                            </div>
              </div>
              <div
                            className="stage"
              >
                            <div
                                          className="progress"
                            >
                                          stage 3
                            </div>
              </div>
</styled.div>,
            <StepField
              value=""
>
               Pelo menos 
              2
               caracteres
</StepField>,
            <StepField
              value=""
>
               Pelo menos 1 letra maiúscula
</StepField>,
            <StepField
              value=""
>
               Pelo menos 1 número
</StepField>,
            false,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Senha",
            },
            "ref": null,
            "rendered": "Senha",
            "type": "label",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "pwd isEmpty",
              "onChange": [Function],
              "placeholder": "",
              "required": true,
              "type": "password",
            },
            "ref": Object {
              "current": null,
            },
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                <div
                  className="stage"
>
                  <div
                                    className="progress"
                  >
                                    stage 1
                  </div>
</div>,
                <div
                  className="stage"
>
                  <div
                                    className="progress"
                  >
                                    stage 2
                  </div>
</div>,
                <div
                  className="stage"
>
                  <div
                                    className="progress"
                  >
                                    stage 3
                  </div>
</div>,
              ],
              "className": "isEmpty",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <div
                    className="progress"
>
                    stage 1
</div>,
                  "className": "stage",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "stage 1",
                    "className": "progress",
                  },
                  "ref": null,
                  "rendered": "stage 1",
                  "type": "div",
                },
                "type": "div",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <div
                    className="progress"
>
                    stage 2
</div>,
                  "className": "stage",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "stage 2",
                    "className": "progress",
                  },
                  "ref": null,
                  "rendered": "stage 2",
                  "type": "div",
                },
                "type": "div",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <div
                    className="progress"
>
                    stage 3
</div>,
                  "className": "stage",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "stage 3",
                    "className": "progress",
                  },
                  "ref": null,
                  "rendered": "stage 3",
                  "type": "div",
                },
                "type": "div",
              },
            ],
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                " Pelo menos ",
                2,
                " caracteres",
              ],
              "value": "",
            },
            "ref": null,
            "rendered": Array [
              " Pelo menos ",
              2,
              " caracteres",
            ],
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": " Pelo menos 1 letra maiúscula",
              "value": "",
            },
            "ref": null,
            "rendered": " Pelo menos 1 letra maiúscula",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": " Pelo menos 1 número",
              "value": "",
            },
            "ref": null,
            "rendered": " Pelo menos 1 número",
            "type": [Function],
          },
          false,
        ],
        "type": [Function],
      },
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
